agent/src/com/cloud/agent/Agent.java

```diff
@@ -990,4 +1091,53 @@ public void doTask(final Task task) throws TaskExecutionException {
             }
         }
     }
+
+    public class HostLBCheckerTask extends ManagedContextTimerTask {
+
+         @Override
+         protected void runInContext() {
+             try {
+                 int index = 0;
+                 final String[] msList = _shell.getHosts();
+                 String preferredHost  = msList[index];
+                 final String connectedHost = _shell.getConnectedHost();
+                 if (s_logger.isTraceEnabled()) {
+                     s_logger.trace("Running host LB checker task, connected host=" + connectedHost + ", preferred host=" + preferredHost);
+                 }
+                 final long wait = _shell.getFailbackPollingWait();
+                 while (!preferredHost.equals(connectedHost)) {
+                     if (_link != null) {
+                         boolean isHostUp = true;
+                         try (final Socket socket = new Socket()) {
+                             socket.connect(new InetSocketAddress(preferredHost, _shell.getPort()), 5000);
+                         } catch (final IOException e) {
+                             isHostUp = false;
+                             if (s_logger.isDebugEnabled()) {
+                                 s_logger.debug("Host: " + preferredHost + " is not reachable, trying to reconnect to next host...");
+                             }
+                         }
+                         if (isHostUp && _link != null) {
+                             long initialTime = System.currentTimeMillis();
+                             while (System.currentTimeMillis() - initialTime < wait && _inProgress.get() > 0) {
```

Comment:
@nvazquez please avoid writing procedural code, in `reconnect` remaining tasks handling already exists. We should not have any timeouts/limits on these tasks, i.e. don't kill/purge outstanding tasks. Instead, one way to rewrite this could be to perform no operations when there are any tasks in `_inProgress`, and let the HostLBCheckerTask retry in the next round.

Reply:
Done, thanks! Agreed